### ISSUE TITLE
No Difference between LongCot-RL and Delethink-RL

### ISSUE THREAD

**Issue by lxianl455** [McGill-NLP/the-markovian-thinker#4]
We tried applying the delethink inference strategy to both officially released models (longcot-1.5b and delethink-1.5b) during the inference phase on the AIME25 dataset and found their performance was roughly the same. This seems to suggest that using delethink during the RL training stage does not bring a clear benefit. 
<img width="1310" height="902" alt="Image" src="https://github.com/user-attachments/assets/86d53918-8c48-4d99-b7a7-426474e6851b" />

Also, the Qwen3 4B Thinking model was trained with LongCot, and their results using the delethink-infer approach are shown in the figure in the paper.

<img width="990" height="531" alt="Image" src="https://github.com/user-attachments/assets/03d67080-29cc-4d19-b05d-bc9bc59c63fa" />

Based on the conclusion above—that for small models there’s no difference between LongCot-rl + delethink-infer and delethink-rl + delethink-infer—it appears that on larger-parameter models, delethink-rl + delethink-infer likewise brings no gains.

I’m not sure whether my understanding is correct.

**Comment by MiladInk** [comment]
Yes. It is correct that delethink-infer applied to R1-Distill 1.5B models trained with longCoT also boosts their performance(in our experiments, delethink achieved near 35% in 128K though which differ from your results a little, can I ask what is your N in avg@N for AIME'25? Also, just to make sure, you're referring to Qwen3-4B in the issue, but the released checkpoints are RL finetunes of R1-Distill 1.5B). **The point is that achieving that LongCoT-RL checkpoints requires quadratic compute while achieving the Delethink-RL checkpoints requires linear compute.** 

I think the points are:
1-Delethink-RL + Delethink-infer => linear compute in training and test-time + test-time scale beyond train-time.
2-LongCoT-RL + LongCoT-infer (the current standard of RL for LLM) => quadratic compute in training and test-time + test-time scale limited to train-time.
3-LongCoT-RL + Delethink-infer => quadratic compute in training, linear compute in test-time + test-time scale beyond train-time

It is true that intellectually the feasibility of bullet 3 is surprising as LongCoT is not trained for Delethink-infer, but I think it is one of strengths of the Delethink that it can zero-shot apply to even LongCoT-RL trained checkpoints (almost all of the current open-source checkpoints). In a sense, this is what makes the performance of both LongCoT-RL and Delethink-RL to match at step zero. Delethink-RL starts with a decent score. 


As a final side note, while zero-shot application of Delethink worked so well on LongCoT-RL trained checkpoint of R1-Distill-1.5B (as we have also reported in our paper), it is not true for all models and markovian state sizes. As shown in Fig F.7 of paper, it works well with C=8K but not with C=16K.

<img width="1030" height="461" alt="Image" src="https://github.com/user-attachments/assets/a45f70cb-0908-4076-b5f8-b8b18bb1fcaf" />

Additionally, it is not true for all models. As evident in "Why Delethink works?", zero-shot application of Delethink does not always boosts original checkpoints performance and Delethink-RL is needed so the model learns to operate with that state size. 

<img width="962" height="474" alt="Image" src="https://github.com/user-attachments/assets/2793c9a1-b10d-43a1-871a-fddd47bf81f3" />

Honestly, I wish it was true that zero-shot application of Delethink could always give a free boost, but it is not always the case. If that was true, we could get higher accuracy, constant memory and linear compute, all with a simple change of inference and no further training. Sometimes it happens, sometimes it does not. But, the presence of markovian samples under the model is consistently observed, which means the RL training starts with many positive samples and has high chance of training well.

**Comment by MiladInk** [comment]
Thanks for sharing your reasoning. First, the most important conclusion of markovian thinker paper is:
**Delethink scales thinking to N tokens with linear compute and constant memory matching scaling of thinking with LongCoT to N tokens.**  

The phenomenon that zero-shot application of Delethink preserves most of the original model's performance is a core discovery of the paper, the fact that both LongCoT-RL 24K  and Delethink-RL 24K runs starts at the same evaluation score, is a positive sign that gives hope that the RL run will succeed. But, the main point is that an RL run with linear compute can match an RL run with quadratic compute. 

However, I think you want evidence for another conclusion: "a model that originally does not support delethink comes to support the delethink strategy after delethink RL training". First, I think the verb "support" here is a bit unclear. Do we want a model that has zero accuracy under zero-shot delethink? The problem is that under Delethink almost all models achieve a respectable zero-shot score. (This is the surprising observation of Delethink paper that off-the-shelf reasoning models contain so much markovian traces. We don't know exactly why, but we guess it is because the thinking processes on the web collected from humans are already very markovian, but this is just mere speculation.)

However, there is evidence to support the conclusion that you're after. "A model with low accuracy under zero-shot Delethink learns to use that small markovian state size well". In the plot below, which I took from the paper, you see that we can reduce the context size to 2K. It drops the base model's performance. The markovian state size is just 1K tokens. But, the model is improving performance over training, surpassing the LongCoT24K performance at initialization. Therefore, we know RL training with smaller markovian sizes is possible.

<img width="1907" height="855" alt="Image" src="https://github.com/user-attachments/assets/aac61afa-6a5b-401f-b400-341a74e7e3bb" />

All in all, the focus of the paper was on showing an alternative to quadratic cost of LongCoT exists without a drop in performance. Therefore, the experiments are geared towards that and the above experiment is a side ablation. I think you're after finding out the exact dynamics that happens in small markovian state sizes, maybe hopefully emergence of summarization "a summarization-moment". I think it is a very interesting direction, but it was not the focus of the paper. We fixed the environment so the model can never exceed its context length, and then it should learn to operate under that limit, and we showed that it matches the LongCoT performance. The thing is this phenomenon of strong zero-shot performance of delethink is so universal, as shown in "Why Delethink works?" that the first problem of studying what you're after is finding a model that does not support delethink, .i.e. its performance drops heavily. Reducing the markovian state size to very small sizes is one way, but to some extent it is not fair: you cannot expect the model to operate on 200 tokens for example. Additionally, I think eyeballing emergence of summarization might not cut it. It might be that the model is subtly carrying important info, not in a explicit summarization paragraph. However, I would not be surprised that in a large RL run this emerges as such actions should be in distribution and can be resurfaced via RL at least. But, these are speculations on my side. 


Regarding the SFT dataset. Have you tried distilling from a LongCoT to sanity check your pipeline?
Are you sampling correct rollouts and then so SFT loss on all chunks? Or, are you SFTing on all rollouts no matter the correctness.
